desc/integrals/bounce_utils.py

```diff
@@ -0,0 +1,850 @@
+
+    """
+    assert w.ndim == 1
+    assert 3 < Q.ndim < 6 and Q.shape[0] == pitch_inv.shape[0] and Q.shape[-1] == w.size
+    assert data["|B|"].shape[-1] == knots.size
+
+    shape = Q.shape
+    Q = Q.reshape(*Q.shape[:3], -1)
+    b_sup_z = interp1d_Hermite_vec(
+        Q,
+        knots,
+        data["B^zeta"] / data["|B|"],
+        data["B^zeta_z|r,a"] / data["|B|"]
+        - data["B^zeta"] * data["|B|_z|r,a"] / data["|B|"] ** 2,
+    )
+    B = interp1d_Hermite_vec(Q, knots, data["|B|"], data["|B|_z|r,a"])
+    # Spline each function separately so that operations in the integrand
+    # that do not preserve smoothness can be captured.
+    f = [interp1d_vec(Q, knots, f_i, method=method) for f_i in f]
+    result = jnp.dot(
+        (
+            integrand(
+                *f,
+                B=B,
+                pitch=1 / pitch_inv[..., jnp.newaxis],
+            )
+            / b_sup_z
+        ).reshape(shape),
+        w,
+    )
+    if check:
+        _check_interp(shape, Q, f, b_sup_z, B, result, plot)
+
+    return result
+
+
+def _check_interp(shape, Q, f, b_sup_z, B, result, plot):
+    """Check for interpolation failures and floating point issues.
+
+    Parameters
+    ----------
+    shape : tuple
+        (P, M, L, Q.shape[-2], w.size).
+    Q : jnp.ndarray
+        Quadrature points in ζ coordinates.
+    f : list[jnp.ndarray]
+        Arguments to the integrand, interpolated to Q.
+    b_sup_z : jnp.ndarray
+        Contravariant toroidal component of magnetic field, interpolated to Q.
+    B : jnp.ndarray
+        Norm of magnetic field, interpolated to Q.
+    result : jnp.ndarray
+        Output of ``_interpolate_and_integrate``.
+    plot : bool
+        Whether to plot stuff.
+
+    """
+    assert jnp.isfinite(Q).all(), "NaN interpolation point."
+    assert not (
+        jnp.isclose(B, 0).any() or jnp.isclose(b_sup_z, 0).any()
+    ), "|B| has vanished, violating the hairy ball theorem."
+
+    # Integrals that we should be computing.
+    marked = jnp.any(Q.reshape(shape) != 0.0, axis=-1)
+    goal = marked.sum()
+
+    assert goal == (marked & jnp.isfinite(b_sup_z).reshape(shape).all(axis=-1)).sum()
+    assert goal == (marked & jnp.isfinite(B).reshape(shape).all(axis=-1)).sum()
+    for f_i in f:
+        assert goal == (marked & jnp.isfinite(f_i).reshape(shape).all(axis=-1)).sum()
+
+    # Number of those integrals that were computed.
+    actual = (marked & jnp.isfinite(result)).sum()
+    assert goal == actual, (
+        f"Lost {goal - actual} integrals from NaN generation in the integrand. This "
+        "is caused by floating point error, usually due to a poor quadrature choice."
+    )
+    if plot:
+        Q = Q.reshape(shape)
+        _plot_check_interp(Q, B.reshape(shape), name=r"$\vert B \vert$")
+        _plot_check_interp(
+            Q, b_sup_z.reshape(shape), name=r"$ (B / \vert B \vert) \cdot e^{\zeta}$"
+        )
+
+
+def _plot_check_interp(Q, V, name=""):
```

Comment:
can you add a test that uses this? seems to be the main source of missing coverage